src/mscorlib/shared/System/Threading/ExecutionContext.cs

```diff
@@ -130,126 +108,164 @@ public static bool IsFlowSuppressed()
+            {
+                // Restore changed ExecutionContext
+                Restore(ref current, executionContext);
+            }
+
+            ExceptionDispatchInfo edi = null;
             try
             {
-                EstablishCopyOnWriteScope(currentThread, ref ecsw);
-                ExecutionContext.Restore(currentThread, executionContext);
-                callback(state);
+                callback.Invoke(state);
             }
-            catch
+            catch (Exception ex)
             {
                 // Note: we have a "catch" rather than a "finally" because we want
                 // to stop the first pass of EH here.  That way we can restore the previous
-                // context before any of our callers' EH filters run.  That means we need to
-                // end the scope separately in the non-exceptional case below.
-                ecsw.Undo(currentThread);
-                throw;
+                // context before any of our callers' EH filters run.
+                edi = ExceptionDispatchInfo.Capture(ex);
             }
-            ecsw.Undo(currentThread);
-        }
-
-        internal static void Restore(Thread currentThread, ExecutionContext executionContext)
-        {
-            Debug.Assert(currentThread == Thread.CurrentThread);
-
-            ExecutionContext previous = currentThread.ExecutionContext ?? Default;
-            currentThread.ExecutionContext = executionContext;
-
-            // New EC could be null if that's what ECS.Undo saved off.
-            // For the purposes of dealing with context change, treat this as the default EC
-            executionContext = executionContext ?? Default;
 
-            if (previous != executionContext)
+            // The common case is that these have not changed, so avoid the cost of a write barrier if not needed.
+            if (currentSyncCtx != previousSyncCtx)
             {
-                OnContextChanged(previous, executionContext);
+                // Restore changed SynchronizationContext back to previous
+                currentSyncCtx = previousSyncCtx;
             }
-        }
 
-        internal static void EstablishCopyOnWriteScope(Thread currentThread, ref ExecutionContextSwitcher ecsw)
-        {
-            Debug.Assert(currentThread == Thread.CurrentThread);
+            if (current != previous)
+            {
+                // Restore changed ExecutionContext back to previous
+                Restore(ref current, previous);
+            }
 
-            ecsw.m_ec = currentThread.ExecutionContext;
-            ecsw.m_sc = currentThread.SynchronizationContext;
+            // If exception was thrown by callback, rethrow it now original contexts are restored
+            edi?.Throw();
         }
 
-        private static void OnContextChanged(ExecutionContext previous, ExecutionContext current)
+        internal static void Restore(ref ExecutionContext current, ExecutionContext next)
```

Comment:
`Restore` seems like the wrong naming for this.  `SetCurrentContext`?

Reply:
Also, maybe `currentRef` would make this easier to understand when reading the code?

Reply:
Changed to `OnValuesChanged` as it only gets triggered if at least once side has notifications